### PR TITLE
NLL is missing struct field suggestion

### PR DESCRIPTION
**Pull request by davidtwco**
Part of #52663.

This commit adds suggestions to change the definitions of fields in
struct definitions from immutable references to mutable references.

r? @nikomatsakis 
cc @pnkfelix

### CODE REVIEW

File: src/librustc_mir/borrow_check/mutability_errors.rs [hunk 1]
```diff
@@ -218,6 +218,31 @@ impl<'a, 'gcx, 'tcx> MirBorrowckCtxt<'a, 'gcx, 'tcx> {
...
+                    elem: ProjectionElem::Field(field, _),
+                }),
+                elem: ProjectionElem::Deref,
+            }) => {
+                err.span_label(span, format!("cannot {ACT}", ACT = act));
+
+                if let Some((span, message)) = annotate_struct_field(
+                    self.infcx.tcx, self.mir, local, field,
+                ) {
+                    err.span_label(span, message);
```
Comment (killercup) [inline src/librustc_mir/borrow_check/mutability_errors.rs]: Could you make this span only the type of the field, and make this a `span_suggestions_with_applicability` with `Applicability::MaybeIncorrect`? That'd make it show up in RLS.

Comment (davidtwco) [reply src/librustc_mir/borrow_check/mutability_errors.rs]: Fixed!

File: src/test/ui/did_you_mean/issue-38147-2.nll.stderr [hunk 2]
```diff
@@ -2,7 +2,7 @@ error[E0596]: cannot borrow `*self.s` as mutable, as it is behind a `&` referenc
   --> $DIR/issue-38147-2.rs:17:9
    |
 LL |     s: &'a String
-   |     ------------- use `&'a mut String` here to make mutable
+   |     ------------- help: consider changing this to be mutable: `&'a mut String`
```
Comment (killercup) [inline src/test/ui/did_you_mean/issue-38147-2.nll.stderr]: The span starts too early: Can you make sure it only covers `&'a String`, as the replacement doesn't contain the `s: ` part?

**Review by killercup — CHANGES_REQUESTED** [review]

File: src/test/ui/did_you_mean/issue-38147-2.nll.stderr (hunk 2 above)
Comment (davidtwco) [reply src/test/ui/did_you_mean/issue-38147-2.nll.stderr]: Fixed!

**Review by killercup — APPROVED** [review]

**Review by nikomatsakis — APPROVED** [review]
r=me but maybe consider matching AST message (could we then remove the `.nll` reference?)

File: src/test/ui/did_you_mean/issue-38147-2.nll.stderr [hunk 3]
```diff
@@ -1,6 +1,9 @@
 error[E0596]: cannot borrow `*self.s` as mutable, as it is behind a `&` reference
   --> $DIR/issue-38147-2.rs:17:9
    |
+LL |     s: &'a String
+   |        ---------- help: consider changing this to be mutable: `&'a mut String`
```
[the remark this answers is not in the document]
Comment (davidtwco) [reply src/test/ui/did_you_mean/issue-38147-2.nll.stderr]: I don't think we can do this. The message did match before this was changed from a label to a suggestion - as a suggestion I think the "change to" part (ie .`&'a mut String`) has to come at the end (or at least, I don't control it), therefore the message needs to change for this to make sense.

[the remark this answers is not in the document]
Comment (killercup) [reply src/test/ui/did_you_mean/issue-38147-2.nll.stderr]: Hm, this is usual diagnostics style for suggestions. I don't think we can easily adjust this "help: {msg}: {replacement}" format.

[the remark this answers is not in the document]
Comment (nikomatsakis) [reply src/test/ui/did_you_mean/issue-38147-2.nll.stderr]: OK

**Review by nikomatsakis — APPROVED** [review]

**Pull request merged**